### ISSUE TITLE
Crash when freeing GrBackendTexture

### ISSUE THREAD

**Issue by DorianRudolph** [rust-skia/rust-skia#750]
I was testing the PR https://github.com/aclysma/skulpin/pull/112 (I have also documented my debugging processes there) and it crashes for me when freeing the GrBackendTexture (`swapchain_destroyed` function in skulpin). This also happens when freeing the GrBackendTexture immediately after creation inside the `VkSkiaSurface::new` function (in skulpin).

The call stack is as follow:
```
___lldb_unnamed_symbol3503 (@___lldb_unnamed_symbol3503:77)
raise (@raise:10)
abort (@abort:50)
___lldb_unnamed_symbol3401 (@___lldb_unnamed_symbol3401:181)
___lldb_unnamed_symbol3567 (@___lldb_unnamed_symbol3567:11)
___lldb_unnamed_symbol3591 (@___lldb_unnamed_symbol3591:293)
__libc_free (@__libc_free:31)
skia_safe::gpu::backend_surface::<impl skia_safe::prelude::NativeDrop for skia_bindings::bindings::GrBackendTexture>::drop (/home/dorian/.cargo/registry/src/github.com-1ecc6299db9ec823/skia-safe-0.57.0/src/gpu/backend_surface.rs:179)
<skia_safe::prelude::Handle<N> as core::ops::drop::Drop>::drop (/home/dorian/.cargo/registry/src/github.com-1ecc6299db9ec823/skia-safe-0.57.0/src/prelude.rs:342)
core::ptr::drop_in_place<skia_safe::prelude::Handle<skia_bindings::bindings::GrBackendTexture>> (@core::ptr::drop_in_place<skia_safe::prelude::Handle<skia_bindings::bindings::GrBackendTexture>>:6)
core::ptr::drop_in_place<skulpin_renderer::skia_support::VkSkiaSurface> (@core::ptr::drop_in_place<skulpin_renderer::skia_support::VkSkiaSurface>:55)
core::ptr::drop_in_place<core::option::Option<skulpin_renderer::skia_support::VkSkiaSurface>> (@core::ptr::drop_in_place<core::option::Option<skulpin_renderer::skia_support::VkSkiaSurface>>:18)
<skulpin_renderer::renderer::SwapchainEventListener as rafx_api::extra::swapchain_helper::RafxSwapchainEventListener>::swapchain_destroyed (/home/dorian/projects/whiteboardcx/skulpin/skulpin-renderer/src/renderer.rs:134)
rafx_api::extra::swapchain_helper::RafxSwapchainHelper::rebuild_swapchain (/home/dorian/.cargo/registry/src/github.com-1ecc6299db9ec823/rafx-api-0.0.14/src/extra/swapchain_helper.rs:492)
rafx_api::extra::swapchain_helper::RafxSwapchainHelper::acquire_next_image (/home/dorian/.cargo/registry/src/github.com-1ecc6299db9ec823/rafx-api-0.0.14/src/extra/swapchain_helper.rs:411)
skulpin_renderer::renderer::Renderer::draw (/home/dorian/projects/whiteboardcx/skulpin/skulpin-renderer/src/renderer.rs:241)
skulpin_app_winit::app::App::run::{{closure}} (/home/dorian/projects/whiteboardcx/skulpin/skulpin-app-winit/src/app.rs:287)
winit::platform_impl::platform::sticky_exit_callback (/home/dorian/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/linux/mod.rs:746)
winit::platform_impl::platform::wayland::event_loop::EventLoop<T>::run_return (/home/dorian/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/linux/wayland/event_loop/mod.rs:376)
winit::platform_impl::platform::wayland::event_loop::EventLoop<T>::run (/home/dorian/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/linux/wayland/event_loop/mod.rs:191)
```

As far as I can tell, the reason is that the GrBackendTexture struct is moved in rust, and thereby copied by value. This struct contain `pub fLabel: std_string`, where `std_string = [u64; 4usize]`. However, in C++ we are not allowed to memcpy `std::string` due to small string optimization (SSO). 

When a string is SSO'd, it contains its own address, which the destructor uses to determine if the string is SSO'd and does not actually need to be freed.
Relevant code from basic_string.h:
```c++
      _GLIBCXX20_CONSTEXPR
      bool
      _M_is_local() const
      { return _M_data() == _M_local_data(); }
      void
      _M_dispose()
      {
	if (!_M_is_local())
	  _M_destroy(_M_allocated_capacity);
      }
      ~basic_string()
      { _M_dispose(); }
```

In [`get_backend_texture`](https://github.com/rust-skia/rust-skia/blob/8eb0dd7fc98dd21e5e8bd2fc866323399b3795f6/skia-safe/src/core/surface.rs#L288), the GrBackendTexture is initialized on the stack. `fLabel` is set to the empty string, so SSO applies. If this GrBackendTexture is now moved by Rust, and subsequently freed by C++, the fLabel string is in an invalid state, as its first 8 bytes point to the original stack address where the GrBackendTexture was first initialized via the placement [new](https://github.com/rust-skia/rust-skia/blob/8eb0dd7fc98dd21e5e8bd2fc866323399b3795f6/skia-bindings/src/gpu.cpp#L130). But then `_M_is_local()` will return false and `_M_destroy` tries to free a stack address, causing a crash.

It seems GrBackendTexture cannot just be treated as an ordinary struct inside Rust.

**Comment by pragmatrix** [comment]
I can reproduce this on Linux with a simple test and with the help of black_box. The solution here is to convert the Rust type to a RefHandle and only allow GrBackendTexture to appear as a heap allocated value on the Rust side. Working on it.

Also note that this does not happen on all platforms. On macOS there does not seem to be a SSO, fLabel is all zero when it represents an empty string.

This was introduced by a recent Skia Milestone update.

**Comment by DorianRudolph** [comment]
Thanks. The difference may be that Mac is using libc++ from clang and Linux
libstdc++ from gcc. Libc++ strings don't store their own address with SSO,
so the empty string is just zeroes ( https://stackoverflow.com/a/21710033)

**Comment by pragmatrix** [comment]
@DorianRudolph I've merged #751 to master, can you test with skulpin from there? Binaries may take a few days.
